### PR TITLE
Allow deployments in pipelines

### PR DESCRIPTION
**Pull request by magdapoppins**
- Add node type "deployment"
- Allow file as edge type

### CODE REVIEW

File: examples/pipeline-example.yaml [hunk 1]
```diff
@@ -42,6 +42,33 @@
...
+            endpoints:
+              - name: predict-digit
+                files:
+                  - name: model
+    edges:
+      #  node.entity_type.entity_name    node.entity_type.entity_name
+      - [batch1.input.training-labels*, train.input.training-labels]
+      - [batch1.metadata.optimal_learning_rate, train.parameter.learning_rate]
+      - [batch1.output.processed-images*, train.input.training-images]
+      - [train.output.model, deploy.file.model]
```
Comment (akx) [inline examples/pipeline-example.yaml:71]: If there are multiple endpoints...
```suggestion
      - [train.output.model, deploy.file.predict-digit.model]
```

File: examples/pipeline-example.yaml [hunk 2]
```diff
@@ -42,6 +42,33 @@
...
+      - name: deploy
+        type: deployment
+        deployment:
+          name: MyDeployment
+          version:
+            name: 2000102.0
+            branch: any
+            commit: none
+            endpoints:
+              - name: predict-digit
```
Comment (akx) [inline examples/pipeline-example.yaml:63]: This must refer to an existing `endpoint:` in the YAML file. This YAML file doesn't seem to be declaring any endpoints?

File: examples/pipeline-example.yaml [hunk 3]
```diff
@@ -42,6 +42,33 @@
...
+                files:
+                  - name: model
```
Comment (akx) [inline examples/pipeline-example.yaml:65]: ... and if it refers to an existing endpoint, that endpoint specifies these, not the pipeline step.
```suggestion
```

File: examples/pipeline-example.yaml [hunk 4]
```diff
@@ -42,6 +42,33 @@
...
+      - name: train
+        type: execution
+        step: Train model
+      - name: deploy
+        type: deployment
+        deployment:
+          name: MyDeployment
+          version:
+            name: 2000102.0
+            branch: any
```
Comment (akx) [inline examples/pipeline-example.yaml:60]: Unnecessary.
```suggestion
```

File: examples/pipeline-example.yaml [hunk 5]
```diff
@@ -42,6 +42,33 @@
...
+            commit: none
```
Comment (akx) [inline examples/pipeline-example.yaml:61]: Why is this required here for deployment nodes if it's not required for execution nodes? (Execution nodes use the commit the blueprint originates from by default; the same should occur here.)

File: examples/pipeline-example.yaml [hunk 6]
```diff
@@ -42,6 +42,33 @@
...
+        step: Batch feature extraction
+      - name: train
+        type: execution
+        step: Train model
+      - name: deploy
+        type: deployment
+        deployment:
+          name: MyDeployment
+          version:
+            name: 2000102.0
```
Comment (akx) [inline examples/pipeline-example.yaml:59]: This can't be static, since attempting to create a version with the same name as a pre-existing version would fail, and there's no such concept as updating a deployment version's configuration; they're very much immutable.

If left empty, we could use the same logic we use for suggesting a new version name in the UI (i.e. generate a `YYYYMMDD.X` version).

I do foresee a need for changing that logic to allow for e.g. a prefix – should we allow some placeholders here? Something like `foo-{datestamp}.{counter}`? (@ruksi, ideas?)

```suggestion
```

File: examples/pipeline-example.yaml [hunk 7]
```diff
@@ -42,6 +42,33 @@
...
+      - name: batch1
+        type: execution
+        step: Batch feature extraction
+      - name: train
+        type: execution
+        step: Train model
+      - name: deploy
+        type: deployment
+        deployment:
+          name: MyDeployment
```
Comment (akx) [inline examples/pipeline-example.yaml:57]: I assume a deployment with this name needs to pre-exist. That's fine.

Comment (akx) [inline examples/pipeline-example.yaml:57]: Considering all the below requests, could a simpler shape for this be

```suggestion
      - name: deploy
        type: deployment
        deployment: MyDeployment
        endpoints:
          - predict-digit
```
or am I missing something?

**Review by akx — CHANGES_REQUESTED** [review]

File: examples/pipeline-example.yaml (hunk 6 above)
Comment (ruksi) [reply examples/pipeline-example.yaml]: we can make the `name` optional and we always add the `{stamp}.{counter}` either way; as the full name or just as suffix

File: valohai_yaml/objs/pipelines/edge.py [hunk 8]
```diff
@@ -7,7 +7,7 @@ def _split_prop(prop: str) -> List[str]:
     return prop.split('.', 2)
 
 
-edge_types = {'input', 'output', 'parameter', 'metadata'}
+edge_types = {'input', 'output', 'parameter', 'metadata', 'files'}
```
Comment (akx) [inline valohai_yaml/objs/pipelines/edge.py:10]:
```suggestion
edge_types = {'input', 'output', 'parameter', 'metadata', 'file'}
```

File: tests/test_pipeline.py [hunk 9]
```diff
@@ -15,6 +15,17 @@ def test_pipeline(pipeline_config: Config):
         )
         for edge in pipeline_config.pipelines["My little pipeline"].edges
     )
+    assert any(
+        (
+            edge.source_node == "train"
+            and edge.source_type == "output"
+            and edge.source_key == "model"
+            and edge.target_node == "deploy"
+            and edge.target_type == "files"
```
Comment (akx) [inline tests/test_pipeline.py:24]:
```suggestion
            and edge.target_type == "file"
```

File: examples/pipeline-example.yaml [hunk 10]
```diff
@@ -42,6 +42,26 @@
...
+        type: deployment
+        deployment: MyDeployment
+        endpoints:
+          - predict-digit
+    edges:
+      #  node.entity_type.entity_name    node.entity_type.entity_name
+      - [batch1.input.training-labels*, train.input.training-labels]
+      - [batch1.metadata.optimal_learning_rate, train.parameter.learning_rate]
+      - [batch1.output.processed-images*, train.input.training-images]
+      - [train.output.model, deploy.files.predict-digit.model]
```
Comment (akx) [inline examples/pipeline-example.yaml:64]:
```suggestion
      - [train.output.model, deploy.file.predict-digit.model]
```

**Review by akx — CHANGES_REQUESTED** [review]
As spoken on the other side, also singularize the edge type here? Other than that I think this is a very good minimal solution.

**Review by akx — APPROVED** [review]
